### PR TITLE
[dagster-airlift] multi-task mapping

### PR DESCRIPTION
**Pull request by dpeng817**
## Summary & Motivation
Allows for mapping the same set of definitions to many dags/tasks. I call a mapping to an individual dag/task a `coupling`. 

An example of how this is going to work:
```python
build_defs_from_airflow_instance(
    instance=instance,
    dag_defs_list=[
        dag_defs(
            "dag1",
            task_defs(
                "task1",
                mark_as_shared("key"),
            ),
        ),
        dag_defs(
            "dag2",
            task_defs(
                "task2",
                mark_as_shared("key"),
            ),
        ),
    ],
    shared_task_defs={"key": Definitions(...)}
)
```

## Changelog
`NOCHANGELOG`

### CODE REVIEW

**Comment by dpeng817** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24464?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#24464** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24464?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#24447** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24447?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24277** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24277?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24321** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24321?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: examples/experimental/dagster-airlift/dagster_airlift/constants.py [hunk 1]
```diff
@@ -1,3 +1,3 @@
 MIGRATED_TAG = "airlift/task_migrated"
-DAG_ID_METADATA_KEY = "airlift/dag_id"
-TASK_ID_METADATA_KEY = "airlift/task_id"
+DAG_ID_METADATA_KEY = "Dag ID"
+DAG_AND_TASK_ID_METADATA_KEY = "airlift/dag_and_task_ids"
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/constants.py:3]: In a follow up, let's change the prefix here to be `dagster-airlift`, the name of the publish pypi package, as that provides a convenient way to namespace tags and metadata in an enforceable way.

**Review by schrockn — COMMENTED** [review]
Let's come up with a noun for (dag_id, task_id) so that we can refer to lists of them.

I'm also skeptical of `RefToDefs` and `SharedSentinel` but I think we need motivating examples to iterate on it.

**Comment by dpeng817** [comment]
@schrockn potential nouns for dag and task id:
`DST (dag scoped task)`
`dagtask`
`coupling`

File: examples/experimental/dagster-airlift/dagster_airlift/constants.py [hunk 2]
```diff
@@ -1,3 +1,6 @@
+from typing import Tuple
+
 MIGRATED_TAG = "airlift/task_migrated"
-DAG_ID_METADATA_KEY = "airlift/dag_id"
-TASK_ID_METADATA_KEY = "airlift/task_id"
+DAG_ID_METADATA_KEY = "Dag ID"
+AIRFLOW_COUPLING_METADATA_KEY = "airlift/couplings"
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/constants.py:5]: Coupling a little odd (and has some understones)

File: examples/experimental/dagster-airlift/dagster_airlift/constants.py [hunk 3]
```diff
@@ -1,3 +1,6 @@
...
+AirflowCoupling = Tuple[str, str]
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/constants.py:6]: Definitely want something with named properties.

File: examples/experimental/dagster-airlift/dagster_airlift/core/dag_defs.py [hunk 4]
```diff
@@ -56,7 +57,51 @@ def assets_def_with_af_metadata(
...
+def mark_as_shared(key: str) -> SharedSentinel:
+    return SharedSentinel(key)
+
+
+class DagDefs:
+    def __init__(self, dag_id: str, task_defs: Dict[str, Union[TaskDefs, RefToDefs]]):
+        self.task_defs = task_defs
+        self.dag_id = dag_id
+
+    def to_defs(self) -> Definitions:
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/dag_defs.py:80]: Need docblock or example usage to understand this

File: examples/experimental/dagster-airlift/dagster_airlift/core/defs_from_airflow.py [hunk 5]
```diff
@@ -17,11 +18,12 @@
 def build_defs_from_airflow_instance(
     airflow_instance: AirflowInstance,
     cache_polling_interval: int = DEFAULT_POLL_INTERVAL,
-    defs: Optional[Definitions] = None,
+    dag_defs_list: List[Union[DagDefs, Definitions]] = [],
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/defs_from_airflow.py:21]: never default to `[]` https://docs.python-guide.org/writing/gotchas/

Comment (schrockn) [reply examples/experimental/dagster-airlift/dagster_airlift/core/defs_from_airflow.py:21]: Seems like we should force the caller to merge all the DagDefs into `Definitions` or _only_ take Dag Defs. This interemediate is not satisfying.

**Review by schrockn — CHANGES_REQUESTED** [review]
We're going to need to talk some of this through. This will definitely have a bunch of downstream consequencs.

**Pull request closed without merging**